active-rfcs/0000-component-tag.md

```diff
@@ -0,0 +1,242 @@
+
+# Example - UI Component
+
+```astro
+<style @component>
+  @import './dep.css'; /* Vite imports supported, including npm packages! */
+  h1 { color: red; }
+</style>
+<script @component>
+  import cowsay from 'cowsay';
+  console.log(cowsay('I am bundled with the rest of the website JS!'));
+</script>
+<h1>Hello!</h1>
+```
+
+# Example - Page or Layout Component
+
+```astro
+<head>
+  <style @component>/* ... */</style>
+</head>
+<body>
+  <script @component>/* ... */</script>
+  <h1>Hello!</h1>
+</body>
+```
+
+
+# Motivation
+
+- Currently, we have an inconsistent API around styling and scripting an Astro component
+  - `<style>` is scoped to the component by default, and you must opt-out via `<style global>`
+  - `<script>` is raw by default, and you must opt-in to bundled behavior (hoisting, uniqueness) via `<script hoist>`
+- **We want** to remove as many inconsistencies as possible in v1.0
+- **We need** bundled scripts and styles to be statically analyzable, for future SSR behavior
+- **We want** any style/script "magic" to be explicit (no default behavior that is confusing to a new user)
+  - This is different from Svelte and Vue, whos API were originally designed only for individual components and not pages.
+- **We need** the ability to directly add inline styles and scripts on the page (ex: add google analytics to page)
+  - This is also something that UI frameworks often struggle with
+
+# Detailed design
+
+This document describes a new API: `@component` . Pronounced as the "Component Tag".
+
+## `@component` General Info
+
+- Identifies a virtual piece of the component that describes the Astro component itself (its look, its behavior, etc)
+  - This is different from every other part of the template, which describes the output HTML.
+- Not available on all types of elements: `<script @component>` and `<style @component>` are the only two supported.
+- Not a new type of directive: `@component` is the only `@` rule that we want to supported.
+- More aggressive than your normal directive: for example, no other attributes are allowed when the `@component` tag is used.
+- Easily statically analyzable by our parser. 
+- Must be easy to statically analyze. Must be top-level or nested inside of a `<head>`/`<body>` element
+  - Compiler error if conditional or nested in any way.
+- Must be static and cannot rely on runtime behavior in any way, so that it can be bundled once.
+  - Compiler error if dynamic in any way (ex: uses `define:vars`).
+
+## `<style>` (no tag)
+
+```astro
+<!-- INPUT: -->
+<style>h1 { color: red; }</style>
+
+<!-- OUTPUT: -->
+<style>h1 { color: red; }</style>
+```
+
+- Style content is always unscoped/global.
+- Style content is escaped as CSS but otherwise untouched by Astro. ex: Sass not supported.
+- `@import` will not be resolved or bundled, sent untouched/raw to the browser.
+- Will be output in the final component render(), useful for adding global CSS in `<head>`.
```

Comment:
Not explicitly said here, but I assume these styles will not be bundled/hoisted, right?

Reply:
thanks! clarified to add this point explicitly